### PR TITLE
Fix bugs in collection

### PR DESCRIPTION
**Pull request by AndreyLykhoman**
*Description*  Fix bugs:
 - Ordering activities after changing type in one activity
 - Ordering after drag and drop
 - Save events (delete, edit...) in activities elements after drug and drop in the collection details  page
 - Save events (delete, edit...) in collections elements after drug and drop in the collection group page
 - Fix problem with close the preview page in the incognito page in Chrom and in Firefox

*Tasks:*
Jira: https://vpalresearch.atlassian.net/browse/AD-262
Youtrack: 
Preview - https://youtrack.raccoongang.com/issue/ALOSI-34
Wrong order - https://youtrack.raccoongang.com/issue/ALOSI-35
Problem with setting in elements - https://youtrack.raccoongang.com/issue/ALOSI-36

### CODE REVIEW

File: bridge_adaptivity/module/views.py [hunk 1]
```diff
@@ -395,6 +403,37 @@ def get(self, request, *args, **kwargs):
...
+    def post(self, request, *args, **kwargs):
+        """
+        To Update activity by a POST request.
+
+        Updating activity and changing the activity's order if activity changed the type.
+        """
+        activity = self.get_object()
+        # NOTE(AndreyLykhoman): 'is_change_type' is the indicator of changing activity type.
+        #    If the 'is_change_type' is True,
+        #    we will have to change the activiry's order.
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: typo s/activiry's/activity's/

File: bridge_adaptivity/module/views.py [hunk 2]
```diff
@@ -395,6 +403,37 @@ def get(self, request, *args, **kwargs):
...
         return super().get(request, *args, **kwargs)
 
+    def post(self, request, *args, **kwargs):
+        """
+        To Update activity by a POST request.
+
+        Updating activity and changing the activity's order if activity changed the type.
+        """
+        activity = self.get_object()
+        # NOTE(AndreyLykhoman): 'is_change_type' is the indicator of changing activity type.
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: typo: s/of changing activity/to change activity/

File: bridge_adaptivity/module/views.py [hunk 3]
```diff
@@ -395,6 +403,37 @@ def get(self, request, *args, **kwargs):
             return HttpResponse(status=201)
         return super().get(request, *args, **kwargs)
 
+    def post(self, request, *args, **kwargs):
+        """
+        To Update activity by a POST request.
+
+        Updating activity and changing the activity's order if activity changed the type.
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: typo: s/if activity changed/if activity changes/

File: bridge_adaptivity/module/views.py [hunk 4]
```diff
@@ -386,7 +386,15 @@ class ActivityUpdate(CollectionSlugToContextMixin, ModalFormMixin, UpdateView):
...
     def get(self, request, *args, **kwargs):
+        """
+        To Update activity by a GET request.
+
+        Updating activities order and running update method
+        in the superclass. The drag and drop feature uses this view.
+        """
         activity = self.get_object()
+        # NOTE(AndreyLykhoman): Changing activity's order
+        #  if kwargs contains the 'order' params.
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: typo: s/params/param/

File: bridge_adaptivity/module/views.py [hunk 5]
```diff
@@ -386,7 +386,15 @@ class ActivityUpdate(CollectionSlugToContextMixin, ModalFormMixin, UpdateView):
...
 
     def get(self, request, *args, **kwargs):
+        """
+        To Update activity by a GET request.
+
+        Updating activities order and running update method
+        in the superclass. The drag and drop feature uses this view.
+        """
         activity = self.get_object()
+        # NOTE(AndreyLykhoman): Changing activity's order
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: ditto: 120 char length is allowed.

File: bridge_adaptivity/module/static/module/js/collection_drag_drop.js [hunk 6]
```diff
@@ -1,13 +1,25 @@
-// include base_drag_drop.js before this script
+// WARNING(AndreyLykhoman):include base_drag_drop.js before this script
 
-function before_add_element(element, index) {
+before_add_element = function(){
```
Comment (idegtiarov) [inline bridge_adaptivity/module/static/module/js/collection_drag_drop.js]: why did you change function declaration to function expression?

please add additional spaces between brackets () and {}

File: bridge_adaptivity/module/static/module/js/collection_drag_drop.js [hunk 7]
```diff
@@ -1,13 +1,25 @@
...
-    label = element.getElementsByClassName("label");
-    if (label.length !== 0){
-        labelItem = label.item(0);
-        labelItem.innerHTML = element.getAttribute("labelstring") + index;
+    // NOTE(AndreyLykhoman): Checking atype of elements and increase index or starting from zero
+    atype = '';
+    index = 0;
+    return function (element) {
+        label = element.getElementsByClassName("label");
+        if (label.length !== 0){
```
Comment (idegtiarov) [inline bridge_adaptivity/module/static/module/js/collection_drag_drop.js]: ditto

File: bridge_adaptivity/module/static/module/js/collection_drag_drop.js [hunk 8]
```diff
@@ -1,13 +1,25 @@
...
+            labelstring = element.getAttribute("data-atype")
+            if (atype !== labelstring){
```
Comment (idegtiarov) [inline bridge_adaptivity/module/static/module/js/collection_drag_drop.js]: ditto

File: bridge_adaptivity/module/views.py [hunk 9]
```diff
@@ -386,7 +386,15 @@ class ActivityUpdate(CollectionSlugToContextMixin, ModalFormMixin, UpdateView):
     context_object_name = 'activity'
 
     def get(self, request, *args, **kwargs):
+        """
+        To Update activity by a GET request.
+
+        Updating activities order and running update method
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: Please use 120 char length limitation

File: bridge_adaptivity/module/views.py [hunk 10]
```diff
@@ -395,6 +403,37 @@ def get(self, request, *args, **kwargs):
...
+        #    If the 'is_change_type' is True,
+        #    we will have to change the activiry's order.
+        is_change_atype = request.POST.get("atype") != activity.atype
+        result = super().post(request, *args, **kwargs)
+        if is_change_atype:
+            # NOTE(AndreyLykhoman): Getting updated activity.
+            activity = self.get_object()
+            ordering_queryset = activity.get_ordering_queryset()
+            ordering_queryset = ordering_queryset.exclude(pk=activity.pk)
+            # NOTE(AndreyLykhoman): if the ordering_queryset contains one
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: ditto

File: bridge_adaptivity/module/views.py [hunk 11]
```diff
@@ -395,6 +403,37 @@ def get(self, request, *args, **kwargs):
...
+            #  or more activities, we can get last element's index,
+            #  increase by one and set to updating activity's order.
+            #  if the ordering_queryset is empty,
+            #  we need to set zero to updating activity's order
+            if ordering_queryset.exists():
+                last = ordering_queryset\
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: what backslash is doing in the code???

File: bridge_adaptivity/module/views.py [hunk 12]
```diff
@@ -646,11 +685,14 @@ def preview_collection(request, slug):
...
     ]
-
     return render(
         request,
         template_name="module/sequence_preview.html",
-        context={'activities': acitvities}
+        context={
+            'activities': acitvities,
+            'back_url': f"{reverse('module:collection-detail', kwargs={'pk':Collection.objects.get(slug=slug).pk})}"
+                        f"?back_url={request.GET.get('back_url')}"
```
Comment (idegtiarov) [inline bridge_adaptivity/module/views.py]: incorrect indentation

File: bridge_adaptivity/module/static/module/js/collection_drag_drop.js (hunk 6 above)
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/static/module/js/collection_drag_drop.js]: 1) This is a higher order function that has two variables in the closure. There are 'atype' and 'index' variables.  The main anonymous function (I call the function right away) returns a function that set to 'before_add_element' variable. And this makes function expression. 
When I use 'before_add_element' for making an element's label, I can check the 'atype' of elements. If the 'atype' of the element isn't equal than variable 'atype', I will set a zero to the 'index' variable, set a new value to 'atype' variable. 
This allows you to make the correct numbering of the elements after drag & drop.

2) I added additional spaces.

File: bridge_adaptivity/module/views.py (hunk 12 above)
Comment (AndreyLykhoman) [reply bridge_adaptivity/module/views.py]: If you meant it should be like this:
```python
        context={
            'activities': acitvities,
            'back_url': f"{reverse('module:collection-detail', kwargs{'pk':Collection.objects.get(slug=slug).pk})}"
                f"?back_url={request.GET.get('back_url')}"
        }
```
The flake8 will rase error:
```
flake8 runtests: commands[0] | flake8
./bridge_adaptivity/module/views.py:689:17: E131 continuation line unaligned for hanging indent
                f"?back_url={request.GET.get('back_url')}"
                ^
ERROR: InvocationError for command '/Volumes/edx/projects/aloci/.tox/flake8/bin/flake8' (exited with code 1)
```

But if I run flake8 without changes, I won't have got any error from flake:
```
flake8 runtests: commands[0] | flake8
______________________________________________________________________________________ summary ______________________________________________________________________________________
  flake8: commands succeeded
  congratulations :)
```

if I didn't undestand you. Please write more information about it.

**Comment by AndreyLykhoman** [comment]
@idegtiarov 
I fixed style mistakes.
